KarhooUISDK/Screens/CheckoutScreen/NewCheckout/Workers/KarhooNewCheckoutBookingWorker.swift

```diff
@@ -77,62 +83,199 @@ final class KarhooNewCheckoutBookingWorker: NewCheckoutBookingWorker {
+                    switch error.type {
+                    case .errMissingBrowserInfo:
+                        self?.sendBookRequest(loyaltyNonce: nil)
+                    default:
+                        self?.bookingState = .failure(error)
+                    }
+                } else if let loyaltyNonce = result.getSuccessValue() {
+                    self?.sendBookRequest(loyaltyNonce: loyaltyNonce.nonce)
+                } else {
+                    self?.sendBookRequest(loyaltyNonce: nil)
+                }
+            }
+        } else {
+            sendBookRequest(loyaltyNonce: nil)
+        }
     }
 
     // MARK: - Booking handling methods
 
-    private func book(paymentNonce: String, passenger: PassengerDetails, flightNumber: String?) {
-        mockBookingSuccess()
+    private func book(paymentNonce: String, passenger: PassengerDetails) {
+        if isLoyaltyEnabled() {
+            loyaltyWorker.getLoyaltyNonce { [weak self] result in
+                if let error = result.getErrorValue() {
+                    if error.type == .errMissingBrowserInfo {
+                        self?.sendBookRequest(loyaltyNonce: nil)
+                    } else {
+                        self?.bookingState = .failure(error)
+                    }
+                    return
+                }
+
+                if let loyaltyNonce = result.getSuccessValue() {
+                    self?.sendBookRequest(loyaltyNonce: loyaltyNonce.nonce)
+                }
+            }
+        } else {
+            sendBookRequest(loyaltyNonce: nil)
+        }
+    }
+
+    func handleThreeDSecureCheck(_ result: OperationResult<ThreeDSecureCheckResult>) {
+        switch result {
+        case .completed(value: let result):
+            switch result {
+            case .failedToInitialisePaymentService:
+                bookingState = .failure(ErrorModel(message: UITexts.PaymentError.noDetailsMessage, code: ""))
+            case .threeDSecureAuthenticationFailed:
+                bookingState = .idle
+                paymentWorker.requestNewPaymentMethod()
+            case .success(let threeDSecureNonce):
+                guard let passengerDetails = passengerDetails else {
+                    assertionFailure()
+                    return
+                }
+                book(paymentNonce: threeDSecureNonce, passenger: passengerDetails)
+            }
+        case .cancelledByUser:
+            bookingState = .idle
+        }
+    }
+
+    private func sendBookRequest(loyaltyNonce: String?) {
+        guard let paymentNonce = paymentWorker.getPaymentNonceAccordingToAuthState(),
+              let passengerDetails = passengerDetails else {
+            bookingState = .failure(ErrorModel(message: UITexts.Errors.somethingWentWrong, code: ""))
+            assertionFailure()
+            return
+        }
+        var flight: String? = flightNumber
+        if let flightText = flight, (flightText.isEmpty || flightText == " ") {
```

Comment:
if we're going with `flightText.isEmpty`, let's go for `flightText.isWhitespace` for consistency and ease of reading